synapse/rest/media/v1/media_repository.py

```diff
@@ -324,14 +321,28 @@ async def _get_remote_media_impl(
 
         # Failed to find the file anywhere, lets download it.
 
-        media_info = await self._download_remote_file(server_name, media_id, file_id)
+        try:
+            media_info = await self._download_remote_file(server_name, media_id,)
+        except SynapseError:
+            raise
+        except Exception as e:
+            # An exception may be because we downloaded media in another
+            # process, so let's check if we magically have the media.
+            media_info = await self.store.get_cached_remote_media(server_name, media_id)
+            if not media_info:
+                raise e
+
+        file_id = media_info["filesystem_id"]
+        file_info = FileInfo(server_name, file_id)
+
+        await self._generate_thumbnails(
+            server_name, media_id, file_id, media_info["media_type"]
+        )
```

Comment:
Shouldn't we only do this when `_download_remote_file` is successful?

If the media magically showed up will this be a waste of effort / conflict / raise an error that we're trying to re-generate thumbnails?

(It looks like this used to be in `download_remote_file`, why was this moved?)

Reply:
I should add a comment, but broadly: 1) it's conceivable that the other download request dies before it generates thumbnails, but mainly b) we want toe be sure the thumbnails have finished being generated before responding to the client, otherwise they'll request thumbnails and get a 404 if they're not ready yet.

Reply:
We moved it out because we want it to get run even if `store_cached_remote_media` throws an duplicate key error

Reply:
Right. So if you have two that are racing and both work, you might get a dup key error for the media, then a dup key error for the thumbnail?

Reply:
We talked through this in #synapse-dev:matrix.org and pretty much we want to ensure the thumbnails exist before existing. So we try it and potentially capture the error in the same way if it already exists. Which is a naive implementation that wastes a bit of work.